### ISSUE TITLE
ScrollToIndex

### ISSUE THREAD

**Issue by exentrich** [jaredLunde/masonic#19]
Huge thanks for such magic library! It deserves way more popularity!
I'm miss one of essential feature is `scrollToIndex`, do you have plans to implement it?
Would be awesome to have something like:
```js
   scrollToIndex({index: 1, align: top, bottom, center, **auto**, behaviour: auto / smooth })
```

**Comment by jaredLunde** [comment]
Hehe, I knew this issue was coming sooner or later 😄 

I've thought about it, but I don't want to add imperative methods to the component. It's more difficult to tell that story in documentation. If I could achieve something with a declarative API (maybe an advanced api feature using a hook) I'd definitely explore that. I've never had a use case for this feature and in general I don't add things to libraries that I'm not getting a personal benefit from. I'm just selfish like that. I know there are people out there who want this feature, though.

I'll poke around at this over the weekend. Something like this seems doable:

```jsx
const {scrollTop, onRender} = useScrollToIndex({scrollTop, index, align, smooth})
return useMasonry({...otherProps, scrollTop, onRender})
```

**Comment by jaredLunde** [comment]
I think I'm circling around this API:

```ts
function useScrollToIndex (
  positioner: Positioner, { 
    offset: number, 
    align: 'center' | 'top' | 'bottom',
    behavior: 'smooth' | 'auto'
  }
): ((index: number) => void)

const App = () => {
  const containerRef = React.useRef(null);
  const [windowWidth, height] = useWindowSize();
  const { offset, width } = useContainerPosition(containerRef, [
    windowWidth,
    height
  ]);
  const positioner = usePositioner({ width, columnGutter: 8 });
  const { scrollTop, isScrolling } = useScroller(offset);
  const scrollToIndex = useScrollToIndex(positioner, {
    offset,
    behavior: "smooth"
  });
  
  // Scroll to index on initial mount
  React.useEffect(() => {
    scrollToIndex(500)
  }, [])

  return useMasonry({
    positioner,
    scrollTop,
    isScrolling,
    height,
    containerRef,
    items: items,
    overscanBy: 3,
    render: FakeCard
  })
};
```

I have a working CodeSandbox so I'm thinking I should have a release for this out by tomorrow.

**Comment by exentrich** [comment]
My use-case is to implement arrow key navigation on items. My app allows user to click an item and make it an "active" so more actions will appear for this item. After user can click Up/Down to navigate further.
Next or previous item can be out of viewport, and there are chance that this item not in DOM right now, so it's imposible to scroll to it. Yes we have an `overscan` option, but it looks dirty to rely on it.

Also I have another use-case: to keep "active" state of an item after page refresh. Scroll to index is crucial in this case.

In the future I plan to add pagination besides infinite scroll, this should improve UX of my app, and this can't be done without `scrollToIndex`.

Any virtualization library have this method, because it's way to hard to implement it outside of a library, because we need to know exact positions of all items, especially ones outside of viewport..

I much appreciate if you make it possible! I like declarative way you suggested, API looks good!
One improvement I want to have is `align = auto`. So I don't need to manually figure out where better to align scroll, and if item is already in viewport - just don't change a scroll position.

**Comment by jaredLunde** [comment]
That's a super valid use case. I'm curious how concerned you are about smooth scrolling. That's really the only thing holding this change up. The positioner is aware of when it has an index's position and it's easy to keep scrolling the window element until that index is found, but with smooth scrolling it becomes either janky or slow depending on the implementation when the best estimate for the index is off. As far as I can tell, there's nothing I can really do about this. One option would be to only enable smooth scrolling when the position of the index is already known, but the UX for that is maybe not ideal because it'd be inconsistent.

Curious to hear your thoughts on this - will definitely have this out soon.

**Comment by jaredLunde** [comment]
Released this in `3.3.0`
